ic-cdk/src/api.rs

```diff
@@ -0,0 +1,497 @@
+///    hooks.
+///  * "canister_update" calls.
+///  * reply or reject callbacks.
+///
+/// # Panics
+///
+/// * This function traps if data.len() > 32.
+/// * This function traps if it's called from an illegal context
+///   (e.g., from a query call).
+pub fn certified_data_set<T: AsRef<[u8]>>(data: T) {
+    let buf = data.as_ref();
+    // SAFETY:  uf is a slice ref, its pointer and length are valid to pass to ic0.certified_data_set.
+    unsafe { ic0::certified_data_set(buf.as_ptr() as usize, buf.len()) }
+}
+
+/// When called from a query call, returns the data certificate authenticating
+/// certified_data set by this canister.
+///
+/// Returns `None` if called not from a query call.
+pub fn data_certificate() -> Option<Vec<u8>> {
+    // SAFETY: ic0.data_certificate_present is always safe to call.
+    if unsafe { ic0::data_certificate_present() } == 0 {
+        return None;
+    }
+    // SAFETY: ic0.data_certificate_size is always safe to call.
+    let n = unsafe { ic0::data_certificate_size() };
+    let mut buf = vec![0u8; n];
+    // SAFETY: Because `buf` is mutable and allocated to `n` bytes, it is valid to receive from ic0.data_certificate_bytes with no offset
+    unsafe {
+        ic0::data_certificate_copy(buf.as_mut_ptr() as usize, 0, n);
+    }
+    Some(buf)
+}
+
+/// Gets current timestamp, in nanoseconds since the epoch (1970-01-01)
+pub fn time() -> u64 {
+    // SAFETY: ic0.time is always safe to call.
+    unsafe { ic0::time() }
+}
+
+/// Sets global timer.
+///
+/// The canister can set a global timer to make the system
+/// schedule a call to the exported `canister_global_timer`
+/// Wasm method after the specified time.
+/// The time must be provided as nanoseconds since 1970-01-01.
+///
+/// The function returns the previous value of the timer.
+/// If no timer is set before invoking the function, then the function returns zero.
+///
+/// Passing zero as an argument to the function deactivates the timer and thus
+/// prevents the system from scheduling calls to the canister's `canister_global_timer` Wasm method.
+pub fn global_timer_set(timestamp: u64) -> u64 {
+    // SAFETY: ic0.global_timer_set is always safe to call.
+    unsafe { ic0::global_timer_set(timestamp) }
+}
+
+/// Gets the value of specified performance counter.
+///
+/// Supported counter types:
+/// * `0` : current execution instruction counter. The number of WebAssembly
+///         instructions the canister has executed since the beginning of the
+///         current Message execution.
+/// * `1` : call context instruction counter. The number of WebAssembly
+///         instructions the canister has executed within the call context
+///         of the current Message execution since Call context creation.
+///         The counter monotonically increases across all message executions
+///         in the call context until the corresponding call context is removed.
+#[inline]
+pub fn performance_counter(counter_type: u32) -> u64 {
```

Comment:
Any reason not to use an enum here and translate the value to the corresponding constants?